ruby/type-hinting-converters-codecs-errors.md

```diff
@@ -0,0 +1,273 @@
+# Ruby SDK - Type hinting, Converters, Codecs and Errors
+
+This proposal goes into details regarding the implementation of data conversion pipeline (including
+type hints, converters and codecs) as well as outlines the hierarchy of exception classes.
+
+## Type hinting
+
+The Ruby typing system is not something we can currently rely on for specifying the types of all the
+inputs and outputs. Instead we will add custom "annotations" (technically these are just method
+calls) to the activity/workflow definitions in order to provide type hinting. Here's the basic
+example:
+
+```ruby
+class SumActivity < Temporal::Activity
+  input Integer, Integer
+  output Integer
+
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+We will support most basic Ruby types (e.g. `Integer`, `Float`, `String`, `Symbol`, `Boolean`,
+`Array`, `Hash`, etc). This will allow the SDK to perform a runtime check of the arguments and
+return values to ensure they match the specified types.
+
+We will also leverage this system to allow SDK users to specify their own types:
+
+```ruby
+class MyInput < Struct.new(:a, :b)
+  def self.load(hash)
+    new(hash['a'], hash['b'])
+  end
+
+  def dump
+    { 'a' => a, 'b' => b } # explicit for illustration purposes, can use #to_h instead
+  end
+end
+
+class MyOutput < Struct.new(:sum)
+  def self.load(sum)
+    new(sum.to_i)
+  end
+
+  def dump
+    sum.to_s # string used for illustration purposes, can use Integer directly instead
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  input MyInput
+  output MyOutput
+
+  def execute(input)
+    MyOutput.new(input.a + input.b)
+  end
+end
+```
+
+Custom types are expected to implement the following interface:
+
+```ruby
+interface _CustomType[U]
+  # Take a primitive type and return an instance of the implementing class
+  def self.load: (U) -> instance
+
+  # Returns a trivially serializable type representing the custom object
+  def dump: -> U
+end
```

Comment:
There's `load`/`dump` methods used by some serialization classes such as `JSON` and `Marshall` (but not user-defined classes). However I don't think there's anything this conflicts with

Reply:
We should use Ruby's and not ours if we can. Ideally there should be nothing Temporal specific about converting classes to/from JSON.

Reply:
But this isn't a converter. We can make this fit into the existing interface (used in `JSON`, `YAML` and `Marshall`), but it won't act as type hinting, more like a custom converter injected an a workflow/activity level

Reply:
To clarify, does Ruby have a way to customize an object for JSON conversion? We should tell people to use that then. Sure we still say _which_ objects, but the "how" is via Ruby conversion.

Reply:
Rails has one (via the `as_json` method that is expected to return an intermediary object that is then easily converted into JSON, like a Hash, Array or String). However we can't depend on Rails, since not everyone is using it.

Ruby allows objects to implement a `to_json` method, but it is expected to generate a JSON String directly (e.g. returning a Hash instead will throw an error).

In both cases there isn't an established way of doing the reverse conversion.

Also, I don't see how JSON is relevant here. This should work with any format, be it YAML, XML or a custom binary.

Reply:
:+1: On the rest of the points.

> Also, I don't see how JSON is relevant here. This should work with any format, be it YAML, XML or a custom binary.

We don't hide the fact that by default we use the language's JSON features, not some "generic" serialization features. In languages like Go and Java, you can customize the object's JSON formatting/parsing via JSON-specific tags/annotations. It is important to those users to know we use stdlib JSON and they can customize in that way.

Reply:
Right, I see, that makes sense 👍 Ruby doesn't have it, so type hints should return something that one of the Payload Converters then can pick up (including their custom converters that they might add).

Reply:
Can you confirm what happens in the JSON payload converter when it sees a class that doesn't have a `dump` (or `load` on decode)?

Reply:
Sure, it then attempts to serialize it using the JSON library. This works well for built-in data types, but obviously fails for custom classes.

One thing to correct — it won't be the JSON payload converter attempting to call `#dump` on a class, this needs to happen before its result is passed to a high-level converter (`Composite` by default). This is done because the result of `dump` might be better converted by using another converter (like `Bytes` or your custom converter)

Reply:
> obviously fails for custom classes.

Not true though, because of https://ruby-doc.org/stdlib-3.0.0/libdoc/json/rdoc/JSON.html#module-JSON-label-Custom+JSON+Additions right?

> One thing to correct — it won't be the JSON payload converter attempting to call #dump on a class, this needs to happen before its result is passed to a high-level converter (Composite by default).

I disagree with this. I think it should only be called at JSON conversion time if present. I don't think the other two converters (bytes and null) should be enough to default to calling dump/load regardless of the child encoding converters. It is up to each encoding converter to handle objects its own way, composite should do nothing special about that.

Reply:
To clarify here, maybe we don't want generic `dump` and `load`, but instead `dump_json` and `load_json` since we shouldn't have generic customization per object, only specific converter customization.

Reply:
@cretz ah, interesting, I wasn't aware of that interface in the JSON class. I haven't ever seen this being used in real world applications, though. Therefore I'm not 100% comfortable depending on this interface.

I also don't want this to only work with JSON. What if I want to use a protobuf here instead without forcing my inputs/outputs to be protobuf generated classes? In this case I would imagine my type hint to return a protobuf object when called `dump` and a protobuf converter would pick it up and encode into a string.

Reply:
Resolved separately